### DISCUSSION TITLE
Are @Scheduled methods already disabled on @Shutdown?

### DISCUSSION THREAD

**Discussion by HerrDerb** [quarkusio/quarkus#38028] [Q&A] [answered]
I have some `@Scheduled` methods for repeated flushing of data to an external system.
In a `@Shutdown` method, I flush the remaining data. Currently, I use a boolean flag `running` which I set to false in my `@Shutdown` method and which is queried within my `@Scheduled` method. This is, because I cannot find any information about the shutdown behaviour of `@Scheduled` methods -> https://quarkus.io/guides/scheduler-reference

simplified:
``` 
@Scheduled(every = "10s")
void schedule(){
  if(running){
    flush();
  }
}

@Shutdown
void schedule(){
 running=false;
 ....
}
```

Is this flag needed at all or are any `@Scheduled` methods already disabled at the time of the shutdown event?

[a comment by an automated account is not shown]

**Comment by mkouba** [comment, marked as answer]
So the `ShutdownEvent` is fired before the CDI container is shut down (otherwise we wouldn't be able to notify the observers) and the scheduler (in both cases, i.e. `quarkus-scheduler` and `quarkus-quartz`) is stopped when the CDI container is shut down. In other words, the `@Scheduled` methods are not "disabled" during the notification.

If you want to get rid of the `running` flag boiler plate then try to use the `io.quarkus.scheduler.Scheduled.ApplicationNotRunning` skip predicate, e.g. something like `@Scheduled(every = "10s", skipExecutionIf = Scheduled.ApplicationNotRunning.class)`.

**Reply by HerrDerb** [reply]
Thank you. `skipExecutionIf = Scheduled.ApplicationNotRunning.class` is an absolut welcome replacement. Maybe this could be added to the Scheduler Refrence Guide?

**Reply by mkouba** [reply]
Sure, why not. Feel free to send a pull request! ;-)

https://quarkus.io/guides/doc-contribute-docs-howto

**Reply by HerrDerb** [reply]
👍 https://github.com/quarkusio/quarkus/pull/38038

**Reply by HerrDerb** [reply]
One follow-up question. Regarding the `ApplicationNotRunning` predicate: https://github.com/quarkusio/quarkus/blob/main/extensions/scheduler/api/src/main/java/io/quarkus/scheduler/Scheduled.java#L268
Would it make sense to explicitly set the priority of the observers? For shutdown = 0 and startup = ~99999 
If no priority defined, what is the default priority? I couldn't find this information, and unluckily it is ommited here: https://quarkus.io/guides/lifecycle#startup_annotation

**Reply by mkouba** [reply]
The default priority of a CDI observer method is `jakarta.enterprise.inject.spi.ObserverMethod.DEFAULT_PRIORITY`, i.e. `2500`. And the same applies to `@Startup`.
